clustermesh: add local endpointslice mirror controller

MrFreezeex opened the pull request:
Please ensure your pull request adheres to the following guidelines:

- [ ] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [ ] Thanks for contributing!

Add a new controller that mirror local EndpointSlice instead of relying on copying the selector of the local Service.

This fixes two issues:
- We were not supporting EndpointSlices not create via selectors which was a known limitation of the existing implementation
- Hostnames were not correctly synchronized with MCS-API. This is because usually hostnames in EndpointSlice are set via a StatefulSet to its governing Services. However in the previous implementation because we effectively change the Service name we were not able to have any hostname set.

This new mirroring controller does fixes those two issues because no selectors are now involved and we fully control the content of the EndpointSlice that we create instead of relying on the controller inside kube-controller-manager.

This new controller uses the suffix of the EndpointSlice as key and mirror the content of two EndpointSlice with the same key.

```release-note
clustermesh: add a new MCS-API controller mirroring the local EndpointSlice instead of copying the Service selectors. Cilium MCS-API now support EndpointSlice created outside of Service selectors and it fixes hostname synchronization in the context of MCS-API.
```

MrFreezeex committed:
clustermesh: add local endpointslice mirror controller

Add a new controller that mirror local EndpointSlice instead of relying
on copying the selector of the local Service.

This fixes two issues:
- We were not supporting EndpointSlices not create via selectors which
  was a known limitation of the existing implementation
- Hostnames were not correctly synchronized with MCS-API. This is
  because usually hostnames in EndpointSlice are set via a StatefulSet
  to its governing Services. However in the previous implementation
  because we effectively change the Service name we were not able to
  have any hostname set.

This new mirroring controller does fixes those two issues because no
selectors are now involved and we fully control the content of the
EndpointSlice that we create instead of relying on the controller inside
kube-controller-manager.

This new controller uses the suffix of the EndpointSlice as key and
mirror the content of two EndpointSlice with the same key.

Signed-off-by: Arthur Outhenin-Chalandre <arthur@cri.epita.fr>

giorio94 commented on pkg/clustermesh/mcsapi/endpointslice_mirror_controller.go:
```diff
@@ -0,0 +1,425 @@
+	)
+	r.Logger.Debug(
+		"MCSAPI EndpointSlice Mirror Reconciler",
+		logAttrReq, req.String(),
```
Nit. Should not call `String()` here, so that it only gets called internally by the logger when debug logs are actually enabled (same for `fmt.Sprint` below).

giorio94 commented on pkg/clustermesh/mcsapi/endpointslice_mirror_controller.go:
```diff
@@ -0,0 +1,425 @@
+		UID:                derivedService.UID,
+		Controller:         ptr.To(true),
+		BlockOwnerDeletion: ptr.To(true),
+	}}
```
Nit. Could use the `controllerutil.SetControllerReference` helper.

giorio94 commented on pkg/clustermesh/mcsapi/endpointslice_mirror_controller.go:
```diff
@@ -0,0 +1,425 @@
+	derivedEndpointSlice.Labels[mcsapiv1alpha1.LabelSourceCluster] = clusterName
+	derivedEndpointSlice.Labels[discoveryv1.LabelManagedBy] = endpointSliceLocalMCSAPIControllerName
+
+	derivedEndpointSlice.Annotations = maps.Clone(derivedService.Annotations)
```
Are service annotations normally propagated to the endpointslices?

MrFreezeex replied:
Ah woops good catch actually no it's only done for labels!

giorio94 commented on pkg/clustermesh/mcsapi/endpointslice_mirror_controller.go:
```diff
@@ -0,0 +1,425 @@
+		if !equalsEndpointPort(&derivedEndpointSlice.Ports[i], &localEndpointSlice.Ports[i]) {
+			return true
+		}
+	}
```
Nit. Could be simplified with `slices.EqualFunc` (same below).

giorio94 commented on pkg/clustermesh/mcsapi/endpointslice_mirror_controller.go:
```diff
@@ -0,0 +1,425 @@
+) error {
+	if len(toAddSuffixes) == 0 {
+		return nil
+	}
```
Nit. Can be dropped for simplicity. Same for the occurrences below.

giorio94 commented on pkg/clustermesh/mcsapi/endpointslice_mirror_controller.go:
```diff
@@ -0,0 +1,425 @@
+	derivedEndpointSlice.Ports = make([]discoveryv1.EndpointPort, 0, len(localEndpointSlice.Ports))
+	for _, port := range localEndpointSlice.Ports {
+		derivedEndpointSlice.Ports = append(derivedEndpointSlice.Ports, *port.DeepCopy())
+	}
```
A shallow copy should be equally fine here as well, given that the resulting object is only potentially mutated during the `Update` call, and we don't access the local endpoints object at that point anymore (controller runtime already returns a deepcopy AFAIR, differently from plain informers/listers).

giorio94 commented on pkg/clustermesh/mcsapi/endpointslice_mirror_controller.go:
```diff
@@ -0,0 +1,425 @@
+			}}}
+		})).
+		Complete(r)
+}
```
One idea to potentially simplify the logic would be to have the controller directly handle epslices, so that the reconciliation unit is a single epslice. The main advantage would be that a change to a given epslice doesn't retrigger the reconciliation of all the other epslices, which can be potentially expensive in case of services with a large number of endpoints (as well as the reconciliation logic is simpler, as 1:1). The reconcile function could then infer the derived service name, check if it exists, and in case proceed with the reconciliation. Additionally, changes in the derived service (only additions actually, as deletions are handled via the owner reference) would need to enqueue a reconciliation for all epslices associated with the corresponding service. Explicitly reacting to non-derived services and serviceexport changes should not be required instead, as they already eventually cause a change in the corresponding derived service via the other controllers. WDYT?

MrFreezeex replied:
We need to react to ServiceExport changes (at least creation and deletion technically) to know if the local service is exported and thus the EndpointSlice need to be mirrored or not. For non derived Services we need to react at least on updating labels and possibly ports (if we start taking it from there). I have never done something like taht through controller-runtime but IIUC I can try to filter precisely to those type of updates to reduce the number of reconciliation.

I am not entirely sure merging that EndpointSlice into one would be simpler as there would be more logic on every reconciliation to merge all the slices together in a sorted way and then check if both slices are equal instead of just comparing the suffix (+ it would be virtually like a normal Endpoint and there is load benefit for the kubernetes apiserver to EndpointSlice vs regular Endpoints)

giorio94 replied:
> We need to react to ServiceExport changes (at least creation and deletion technically) to know if the local service is exported and thus the EndpointSlice need to be mirrored or not.

Ah, right, because we may have a derived service if any of the remote clusters is exporting their service, even if the local one is not exported. 

> For non derived Services we need to react at least on updating labels and possibly ports (if we start taking it from there).

Aren't we copying the labels from the derived service?

>  I have never done something like taht through controller-runtime but IIUC I can try to filter precisely to those type of updates to reduce the number of reconciliation.

That should be possible by using ` handler.TypedFuncs` instead of `handler.EnqueueRequestsFromMapFunc`, and implementing only the methods that you are interested in. Not sure whether it is worthy going into that extra complexity at the moment though.

> I am not entirely sure merging that EndpointSlice into one would be simpler

Ah, sorry, that's not what I meant (and I agree with all your points there). What I was suggesting is having the reconciliation operate at the single epslice level (i.e., the key of reconcile be the nsname of the epslice, rather than the service export one). Given that all epslices are independent, you can reconcile them one by one, rather than in a batch, potentially also saving a few operations if only one gets modified.

MrFreezeex replied:
>>   For non derived Services we need to react at least on updating labels and possibly ports (if we start taking it from there).

> Aren't we copying the labels from the derived service?

Ah yep indeed :man_facepalming:, so yes it look like we don't need to reconcile there!

> Ah, sorry, that's not what I meant (and I agree with all your points there). What I was suggesting is having the reconciliation operate at the single epslice level (i.e., the key of reconcile be the nsname of the epslice, rather than the service export one). Given that all epslices are independent, you can reconcile them one by one, rather than in a batch, potentially also saving a few operations if only one gets modified.

Ah yes indeed! will look into this!

MrFreezeex replied:
Ok so I started the version that reconcile EndpointSlice individually here: https://github.com/MrFreezeex/cilium/blob/mirror-endpointslices-indiv/pkg/clustermesh/mcsapi/endpointslice_mirror_controller.go. This not done yet and not tested (I didn't updated the test + didn't tested on a cluster too) but the logic should be mostly there I think (there is probably a few bugs/typo though).

So the code is actually longer perhaps or equivalent in term of LOC... Basically while our Reconcile method is less complex, there is more things to do in the watchers logic. I think I slightly prefer reconciling EndpointSlice individually still but would love to have your opinion on which version I should continue working on.

giorio94 replied:
Sorry for the delay, this slipped under my radar. Yeah, the new version overall makes sense to me, it seems more straightforward to reason about. It shouldn't be also difficult to reduce a bit the overall length by trimming some duplication here and there, but that's definitely not the most important part, and we can discuss about that during the review.

giorio94 commented on pkg/clustermesh/mcsapi/endpointslice_mirror_controller.go:
```diff
@@ -0,0 +1,425 @@
+// getSuffix return the name of the EndpointSlice trimmed by its generatedName
+func getSuffix(endpointSlice *discoveryv1.EndpointSlice) string {
+	return strings.TrimPrefix(endpointSlice.Name, endpointSlice.GenerateName)
+}
```
I would suggest adding a dedicated annotation on the derived epslices with the name of the source epslice, rather than relying on the generate name here, just to prevent surprises in case, say, the epslice was originally created without generate name (i.e., by some external tooling).

MrFreezeex replied:
We control the target EndpointSlice where we already force a `GenerateName` so adding a new annotation there won't really change the behavior too much here. It might be clearer where the EndpointSlice came from for a user reading its content though.

Currently if the EndpointSlice is created manually/by some external tooling without GenerateName it would essentially take the whole name as suffix, which should work most of the time (except if the name is really long).

What we could do to alleviate this behavior could be to also trim the service name from the EndpointSlice name, although if the EndpointSlice is created with some very unrelated name it would still take the full name. And if we want to address the possible too long name we can try to take the last like 40 characters of the suffix that we detect (the derived named is always 18 characters and I think we should not go over 63 IIRC).

giorio94 replied:
> We control the target EndpointSlice where we already force a GenerateName so adding a new annotation there won't really change the behavior too much here. It might be clearer where the EndpointSlice came from for a user reading its content though.

Yeah, it feels hacky to me setting the GenerateName but then forcing the Name, just for the side effect of then inferring the link to the corresponding original epslice. Having a dedicated annotation would be a bit more explicit instead (or alternatively an owner reference, but that would change the logic a bit). I don't feel this a hard blocker though.

> Currently if the EndpointSlice is created manually/by some external tooling without GenerateName it would essentially take the whole name as suffix, which should work most of the time (except if the name is really long).

> What we could do to alleviate this behavior could be to also trim the service name from the EndpointSlice name, although if the EndpointSlice is created with some very unrelated name it would still take the full name. And if we want to address the possible too long name we can try to take the last like 40 characters of the suffix that we detect (the derived named is always 18 characters and I think we should not go over 63 IIRC).

Yep, there are a few caveats here and there. We could just use a generate name (and then use a label to associate the original epslice), but that makes things a bit more complex due to the requirement of internal tracking to avoid creating multiple derived epslices for the same original one due to the previously created object not having been received by the informer yet. But in general, yes, trimming the suffix seems a safe approach to prevent most of the problems.

MrFreezeex replied:
> Yeah, it feels hacky to me setting the GenerateName but then forcing the Name, just for the side effect of then inferring the link to the corresponding original epslice. Having a dedicated annotation would be a bit more explicit instead (or alternatively an owner reference, but that would change the logic a bit). I don't feel this a hard blocker though.

I separated that logic change in its own commit, if you think that's better we can just rely on the service label and not setting the generateName at all instead, we would still be able to apply the same logic to all the EndpointSlice whether they are derived or not

giorio94 commented on pkg/clustermesh/mcsapi/endpointslice_mirror_controller.go:
```diff
@@ -0,0 +1,425 @@
+	logAttrToUpdateSuffixes = "toUpdateSuffixes"
+	logAttrToAddSuffixes    = "toAddSuffixes"
+	logAttrToDeleteSuffixes = "toDeleteSuffixes"
+)
```
Nit. Most of the log attributes are currently defined in `pkg/logging/logfields.go`. It could make sense moving these there as well, to favor consistency and reuse.

MrFreezeex replied:
The "req" is already there that's true, for the other three should I add it there even though it's kind off specific to this controller?

giorio94 replied:
Both approaches should work, I agree they are difficult to reuse.

giorio94 reviewed (commented):
Thanks! I've started taking a first look, and left a few suggestions/questions inline.

One additional aspect that may be worth considering is that the derived service may not exactly match the original service, both due to conflict resolution, and because of address family differences (if I'm not mistaken, derived services are always IPv4 only, which is a suboptimal). Hence, plainly copying the epslices may not necessarily be strictly correct in all cases. While conflicts (only ports should affect the behavior) may be considered niche enough to be neglected, address family behavior may need additional considerations.

giorio94 commented on pkg/clustermesh/mcsapi/endpointslice_mirror_controller.go:
```diff
@@ -0,0 +1,477 @@
+				requests = append(requests, ctrl.Request{NamespacedName: types.NamespacedName{
+					Name:      epSlice.Name,
+					Namespace: epSlice.Namespace,
+				}})
```
Nit. 

```suggestion
				requests = append(requests, ctrl.Request{NamespacedName: client.ObjectKeyFromObject(&epSlice)})
```

giorio94 commented on pkg/clustermesh/mcsapi/endpointslice_mirror_controller.go:
```diff
@@ -0,0 +1,477 @@
+func getSuffix(endpointSlice *discoveryv1.EndpointSlice) string {
+	suffix := strings.TrimPrefix(endpointSlice.Name, endpointSlice.Labels[discoveryv1.LabelServiceName]+"-")
+	suffixLen := min(40, len(suffix))
+	suffix = strings.TrimPrefix(suffix[len(suffix)-suffixLen:], "-")
```
Technically endpointslice names can also contain dots, which should be sanitized as well.

```suggestion
	suffix = strings.TrimLeft(suffix[len(suffix)-suffixLen:], "-.")
```

giorio94 commented on pkg/clustermesh/mcsapi/endpointslice_mirror_controller.go:
```diff
@@ -0,0 +1,477 @@
+						Namespace: epSlice.Namespace,
+					}})
+				}
+				return requests
```
Nit. This looks fairly identical to the one above (besides the label value). I'd suggest extracting it into a separate function to reduce duplication.

giorio94 commented on pkg/clustermesh/mcsapi/endpointslice_mirror_controller.go:
```diff
@@ -0,0 +1,477 @@
+				Namespace: derivedEpSlice.Namespace,
+			},
+		)
+		if err != nil || derivedService == nil {
```
Nit. Rather than checking `derivedService == nil`, I would propagate the NotFound error here as well (directly calling`client.Get` to get rid of the helper) , and then ignore it below. IMO that makes the intent a bit clearer.

giorio94 commented on pkg/clustermesh/mcsapi/endpointslice_mirror_controller.go:
```diff
@@ -0,0 +1,477 @@
+func getLocalDerivedEndpointSliceName(localEpSlice *discoveryv1.EndpointSlice) string {
+	suffix := getSuffix(localEpSlice)
+	derivedServiceName := getDerivedServiceName(localEpSlice)
+	if derivedServiceName == "" || suffix == "" {
```
AFAICT the suffix cannot ever be empty, because an object name must always end with an alphanumeric character: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-subdomain-names

MrFreezeex replied:
It can be assuming that someone decide to create an EndpointSlice with the exact same same as its Service manually, although I guess I should just properly handle that case and not fail midway through though

giorio94 replied:
There will not be problems even in that case, as you are trimming `<service-name>-`, so if the epslice name matches that of the service nothing would be trimmed. And you are guaranteed that the remaining part will always be a valid name (if it is trimmed) because a name must end with an alphanumeric character.

MrFreezeex replied:
About this I am now trimming `<service-name>` instead of `<service-name>-` and handling the empty suffix without throwing an error

giorio94 commented on pkg/clustermesh/mcsapi/endpointslice_mirror_controller.go:
```diff
@@ -0,0 +1,477 @@
+	}
+	if err != nil || derivedService == nil {
+		return controllerruntime.Fail(err)
+	}
```
Nit. I would move this to the two branches, just to make the logic more clear.

giorio94 commented on pkg/clustermesh/mcsapi/endpointslice_mirror_controller.go:
```diff
@@ -0,0 +1,477 @@
+) {
+	controllerutil.SetControllerReference(derivedService, derivedEpSlice, r.Scheme())
+
+	derivedEpSlice.Labels = maps.Clone(derivedService.Labels)
```
This overwrites all existing labels, which may not be desired (it can create problems in case another controller adds its own labels). I would replace it with `maps.Copy(derivedEpSlice.Labels, derivedService.Labels)`.

MrFreezeex replied:
Hmmm this looks very interesting but this in a bit in a weird spot since most of the time labels are set this way if generated from pods via the  service selector and the KCM (https://github.com/cilium/endpointslice/blob/2acdeef56068cc3b78899ceb100f1bf7955dc4ad/utils.go#L217). If something else than the KCM generate their EndpointSlice from scratch though they can add their own labels to regular EndpointSlice.

However since we also don't currently mirror those to remote clusters I would say that we shouldn't allow that for now for global consistency. If EndpointSlice sync starts to also mirror labels (same for annotations also) we should be able to allow some of those custom labels/annotations though.

Also assuming we start to sync EndpointSlice labels/annotations to remote clusters I am wondering if we should try to track "custom" annotations/labels (not inherited from the local Service) and propagate that instead. Because if we allow direct custom label/annotation to the derived EndpointSlice it would be hard to do inherited removal of labels/annotations from derived service :/.

So to me this would be more a topic for later/low prio and perhaps even a recommendation to be integrated in the KEP directly rather than deciding what to do locally for Cilium since it's not obvious what should be the correct behavior anyway...

giorio94 replied:
Sorry if I haven't been clear in my previous comment.

I wasn't referring at all at propagating the labels/annotations to remote clusters --- I cannot think of any valid use-case for that, as they don't have a specific meaning, and users are not expected to interact directly with epslices.

My only concern was preventing an endless reconciliation loop if for whatever reason there were to be a controller/webhook that, say, adds a custom annotation. I know all of this is kind of niche, but I've been bit in the past by this kind of *controller fights* due to too strict reconciliation policies.

That said, I've experimented a bit more with how the K8s controller handles the update of labels and annotations, and it apparently enforces the labels (reverting user-chances), while annotations are preserved. That seems fairly easy to implement (just pass through existing ones, and enforce the ones coming from the reflected object --- without taking care of deletions), and it would make the logic sort of consistent with the upstream one. WDYT?

MrFreezeex replied:
> Sorry if I haven't been clear in my previous comment.
> 
> I wasn't referring at all at propagating the labels/annotations to remote clusters --- I cannot think of any valid use-case for that, as they don't have a specific meaning, and users are not expected to interact directly with epslices.

Sorry I got that but but I wasn't probably so clear too, my point was that with MCS-API we should tend to have as much as possible a "global consistency". So essentially if there's a valid use case for modifying labels/annotations of local epslice (meaning either the derived or non derived one somehow as long as it's not on the remote epslice ones)  those should also be propagated to other clusters somehow for global consistency. But while I am not a big fan of doing that we could still allow updates to annotations to match regular generated EndpointSlice but I don't think we should allow it for labels for now though.

giorio94 replied:
> So essentially if there's a valid use case for modifying labels/annotations of local epslice [...] those should also be propagated to other clusters somehow for global consistency.

I don't personally agree here, because IMO the epslice annotation is a local property which is not expected to influence anything (conversely to service annotations, which are often used for LB purposes). Trying to reconcile them globally would be extremely painful (endpoints are not guaranteed to be grouped in the same slice in different clusters), with most likely no practical added value. 

>  we could still allow updates to annotations to match regular generated EndpointSlice but I don't think we should allow it for labels for now though.

Agreed about not allowing changes for labels, the k8s controller-manager itself doesn't seem to allow that. As for annotations, I think the most consistent approach would be simply preserving the ones that a user may set on the derived object, rather than explicitly removing them as performed currently.

MrFreezeex replied:
> Trying to reconcile them globally would be extremely painful (endpoints are not guaranteed to be grouped in the same slice in different clusters), with most likely no practical added value.

Ouch yes you are right I forgot about the remote epslice grouping!

giorio94 commented on pkg/clustermesh/mcsapi/endpointslice_mirror_controller.go:
```diff
@@ -0,0 +1,477 @@
+	derivedEpSlice.Labels[mcsapiv1alpha1.LabelSourceCluster] = r.clusterName
+	derivedEpSlice.Labels[discoveryv1.LabelManagedBy] = endpointSliceLocalMCSAPIControllerName
+
+	derivedEpSlice.Annotations = nil
```
I would not delete existing annotations, for the same reason.

giorio94 commented on pkg/clustermesh/mcsapi/endpointslice_mirror_controller.go:
```diff
@@ -0,0 +1,477 @@
+) bool {
+	desiredDerivedEndpointSlice := r.newDerivedEndpointSlice(localEpSlice, derivedService)
+
+	if !maps.Equal(derivedEpSlice.Labels, desiredDerivedEndpointSlice.Labels) {
+		return true
+	}
```
If user labels are preserved, this should be changed to only assert that all expected labels are present.

giorio94 commented on pkg/clustermesh/mcsapi/endpointslice_mirror_controller.go:
```diff
@@ -0,0 +1,477 @@
+	}
+	if !maps.Equal(derivedEpSlice.Annotations, desiredDerivedEndpointSlice.Annotations) {
+		return true
+	}
```
If user annotations are preserved, this should be dropped.

giorio94 commented on pkg/clustermesh/mcsapi/endpointslice_mirror_controller.go:
```diff
@@ -0,0 +1,477 @@
+	if !slices.EqualFunc(derivedEpSlice.Endpoints, localEpSlice.Endpoints, equalsEndpoint) {
+		return true
+	}
+	desiredPorts := shallowCopyServicePortToEndpointPort(derivedService.Spec.Ports)
```
Unless I'm missing something, this appears redundant, as already computed by `r.newDerivedEndpointSlice`.

giorio94 commented on pkg/clustermesh/mcsapi/endpointslice_mirror_controller.go:
```diff
@@ -0,0 +1,477 @@
+	return true
+}
+
+func equalsEndpoint(a, b discoveryv1.Endpoint) bool {
```
I wonder if `reflect.DeepEqual` could be a solution here (although probably a bit more inefficient), preventing issues in case any new field gets added in the future. Same above.

giorio94 commented on pkg/clustermesh/mcsapi/endpointslice_mirror_controller.go:
```diff
@@ -0,0 +1,477 @@
+}
+
+func (r *mcsAPIEndpointSliceMirrorReconciler) getLocalEndpointSlice(
+	ctx context.Context, derivedEpSlice *discoveryv1.EndpointSlice, derivedService *corev1.Service,
```
Nit. `derivedService` is unused.

giorio94 commented on pkg/clustermesh/mcsapi/endpointslice_mirror_controller.go:
```diff
@@ -0,0 +1,477 @@
+		derivedService, err = r.getDerivedService(
+			ctx,
+			types.NamespacedName{
+				Name:      derivedEpSlice.Labels[discoveryv1.LabelServiceName],
```
The handling of the `discoveryv1.LabelServiceName` label looks a bit inconsistent to me, as sometimes you explicitly check for its presence, and sometimes not. How about checking for its presence immediately at the beginning of the `Reconcile` function (and the watch handler functions), and returning if not set (as if the object was not found)? That would allow to avoid the checks deep down in the call stack, which can be confusing, as at the end of the day there is not much that can be done.

MrFreezeex replied:
I will double check what I can do to improve that however it isn't as obvious because I can't treat exactly as if the object was not found for derived EndpointSlice because then the controller might try to do a create call while it still exist which means if someone decide to remove some of those labels manually the controller will be essentially stuck :/...

giorio94 replied:
I think it is totally acceptable to not reconcile the object if the label has been manually removed, as long as the reconciliation process does not get stuck in an endless loop. And I agree that this case is tricky (if not impossible) to handle --- I would just make the behavior explicit, rather than performing checks deep down in the call stack, which are difficult to reason about.

MrFreezeex replied:
In this new version I tried to focus consolidating these into some helper function so while I am not doing it early I am doing this in a mostly central place, let me know wdyt I can totally add some more early check too

giorio94 commented on pkg/clustermesh/mcsapi/endpointslice_mirror_controller_test.go:
```diff
@@ -0,0 +1,363 @@
+				Name:      "full-" + suffix,
+				Namespace: "default",
+			}
+			result, err := r.Reconcile(context.Background(), ctrl.Request{
```
Nit. `s/context.Background()/t.Context()/` (same for all occurrences).

giorio94 commented on pkg/clustermesh/mcsapi/endpointslice_mirror_controller.go:
```diff
@@ -0,0 +1,477 @@
+	endpointPorts := make([]discoveryv1.EndpointPort, 0, len(ports))
+	for _, port := range ports {
+		endpointPort := discoveryv1.EndpointPort{
+			Port:        &port.Port,
```
This should be the target port, not port.

MrFreezeex replied:
Argh yes... So if we want to keep this somewhat simple then the ports should unfortunately not be taken out of service then because it means that we would have to translate a potential named port to int, aka watching for pods too...

Only mirroring ports already existing on the local EndpointSlice still make sense in a way I think it "just" means that user querying a port would only land on Endpoints where these ports are properly defined rather than every endpoints.

giorio94 replied:
Yeah, preserving the ones from the local epslice seems the easiest to avoid having to reimplement a lot of conversion logic. I'm not exactly sure if there could be cases of conflicts that cause a different subset of ports to be chosen by the service (same name but different port maybe?), but handling those is probably out of scope here, at least for the moment.

MrFreezeex replied:
Hmm yep also I think most of those weird cases should be caught by a ServiceExport conflict, then the behavior might not be the best on port conflict but at least there is some kind of warning exposed to tell the user that there is something wrong...

giorio94 commented on pkg/clustermesh/mcsapi/endpointslice_mirror_controller.go:
```diff
@@ -0,0 +1,477 @@
+	var localEpSlice discoveryv1.EndpointSlice
+	if err := r.Client.Get(
+		ctx,
+		types.NamespacedName{Name: serviceName + "-" + suffix, Namespace: derivedEpSlice.Namespace},
```
This breaks if the local epslice is not named after the service name. A more robust approach could be having an annotation in the derived epslice with the name of the corresponding original one.

giorio94 reviewed (commented):
Thanks, and sorry for the delay. 

Overall the structure looks good to me, I've left a few more comments inline, mostly nits and a couple of suggestions to make the logic more robust in corner cases.

FYI, the new "script based" testing framework may be neat in this context (no need to change the tests of this PR at the moment though). You can find an example [here](https://github.com/cilium/cilium/blob/27ed72467d006d818affe936df01c77abfd2d4e0/pkg/loadbalancer/tests), but basically the idea is having a test that sets up the entire set of components, and then defining the actual test cases by means of self-contained script files (`txtar`).

giorio94 commented on pkg/clustermesh/mcsapi/endpointslice_mirror_controller.go:
```diff
@@ -0,0 +1,399 @@
+		Watches(&corev1.Service{}, handler.EnqueueRequestsFromMapFunc(func(ctx context.Context, obj client.Object) []ctrl.Request {
+			svcImportOwner := getOwnerReferenceName(obj.GetOwnerReferences(), mcsapiv1alpha1.GroupVersion.String(), kindServiceImport)
+			if svcImportOwner != "" {
+				return r.getEndpointSliceFromServiceRequests(ctx, types.NamespacedName{Name: kindServiceImport, Namespace: obj.GetNamespace()})
```
I assume name should be `svcImportOwner`.

giorio94 commented on pkg/clustermesh/mcsapi/endpointslice_mirror_controller.go:
```diff
@@ -0,0 +1,399 @@
+		return []ctrl.Request{}
+	}
+
+	requests := []ctrl.Request{}
```
Nit. You can already initialize the slice of the correct length.

giorio94 commented on pkg/clustermesh/mcsapi/endpointslice_mirror_controller.go:
```diff
@@ -0,0 +1,399 @@
+
+	var epSliceList discoveryv1.EndpointSliceList
+	if err := r.Client.List(ctx, &epSliceList, &client.ListOptions{Namespace: key.Namespace, LabelSelector: selector}); err != nil {
+		return []ctrl.Request{}
```
Nit. `return nil` would save an unnecessary allocation (same above).

giorio94 commented on pkg/clustermesh/mcsapi/endpointslice_mirror_controller.go:
```diff
@@ -0,0 +1,399 @@
+		if isExported, err := r.isServiceExported(
+			ctx,
+			types.NamespacedName{
+				Name:      localEpSlice.Labels[discoveryv1.LabelServiceName],
```
Should likely handle the fact that the label may not be present, as the name would then be nil, which I assume would cause the client to trigger an error. Hence, entering into a backoff loop which is unnecessary.

giorio94 commented on pkg/clustermesh/mcsapi/endpointslice_mirror_controller.go:
```diff
@@ -0,0 +1,399 @@
+	}
+
+	if derivedServiceName == "" {
+		return controllerruntime.Success()
```
Could be worth adding a comment on the reason for returning here, and maybe also emitting a log line just to simplify investigating possible unexpected behavior.

MrFreezeex replied:
Ok soooo I went ahead and tried to handle those as besides the user fiddling around the mirrored endpointslice which could be fair that we do not handle (I still handled to the best I could though), there is actually a valid (but niche) case where the EndpointSlice is manually created and the user/something decide to remove the service label name instead of deleting it...

giorio94 commented on pkg/clustermesh/mcsapi/endpointslice_mirror_controller.go:
```diff
@@ -0,0 +1,399 @@
+		derivedServiceName = derivedEpSlice.Labels[discoveryv1.LabelServiceName]
+	} else {
+		localEpSlice = &epSlice
+		if isExported, err := r.isServiceExported(
```
I wonder if it would be better to simply drop this check here, in favor of the one in `shouldMirrorLocalEndpointSlice`. WDYT?

MrFreezeex replied:
Yes actually I think that would be right, will check to simplify that thanks! :eyes:

giorio94 commented on pkg/clustermesh/mcsapi/endpointslice_mirror_controller.go:
```diff
@@ -0,0 +1,399 @@
+	derivedEpSliceKey := getLocalDerivedEndpointSliceKey(localEpSlice)
+	if derivedEpSliceKey == nil {
+		return nil
+	}
```
Things may likely go wrong if nil is returned here. I reckon that it can never happen, because you already validated that `derivedServiceName` is not empty. How about adding a clarification comment, and dropping this return?

giorio94 reviewed (commented):
Thanks! A couple of minor comments inline, before performing the final test.

MrFreezeex committed:
clustermesh: handle tricky case in the endpointslice mirror controller

This commit attempt to recover nicely on tricky cases where the service
name label is remove from EndpointSlice.

Signed-off-by: Arthur Outhenin-Chalandre <arthur@cri.epita.fr>

giorio94 commented on pkg/clustermesh/mcsapi/endpointslice_mirror_controller.go:
```diff
@@ -0,0 +1,447 @@
+				)
+				return r.Reconcile(ctx, ctrl.Request{
+					NamespacedName: client.ObjectKeyFromObject(localEpSlice),
+				})
```
Manually calling `Reconcile` is a bit risky, because it breaks the assumption that `Reconcile` is never called in parallel for the same key. It is probably not a big deal in this case, as there should be a single worker nonetheless. Additionally, the retry in case of an error may not necessarily work, depending on whether the local or the derived epslice triggered the original reconciliation. 

That said, it should not be even needed to explicitly call reconcile here, because the deletion of the derived epslice would trigger an event that enqueues the local epslice anyways.

MrFreezeex replied:
Ah yes I didn't thought about the retry and indeed it should be already tackled by the deletion watcher indeed 👀

giorio94 reviewed (approved):
Thanks! One more comment inline. Additionally, the cilium operator's cluster role should be also updated to allow endpointslice write operations when MCS-API is enabled, as this controller would otherwise fail due to lack of permissions if epslice synchronization is disabled.
